### PR TITLE
API Add a CREATE_MEMORY_HYDRATED option to DataObject constructor

### PR DESCRIPTION
**Pull request by maxime-rainville**
This adds a fourth option to the DataObject constructor to hydrate from memory.

Fixes https://github.com/silverstripe/silverstripe-framework/issues/9750

### CODE REVIEW

File: src/ORM/DataObject.php [hunk 1]
```diff
@@ -203,6 +203,12 @@ class DataObject extends ViewableData implements DataObjectInterface, i18nEntity
      */
     const CREATE_HYDRATED = 2;
 
+    /**
+     * Value for 2nd argument to constructor, indicating that a record is being hydrated from memory. This can be used
+     * to initialised a record that doesn't yet have an ID. Neither setters and nor default population will be called.
```
Comment (madmatt) [inline src/ORM/DataObject.php:208]: Sentence could be structured a bit more clearly - e.g. `Setter methods are not called, and population via private static $defaults will not occur`.

File: src/ORM/DataObject.php [hunk 2]
```diff
@@ -339,7 +345,9 @@ public static function getSchema()
      * Construct a new DataObject.
      *
      * @param array $record Initial record content, or rehydrated record content, depending on $creationType
-     * @param int|boolean $creationType Set to DataObject::CREATE_OBJECT, DataObject::CREATE_HYDRATED, or DataObject::CREATE_SINGLETON. Used by SilverStripe internals as best left as the default by regular users.
+     * @param int|boolean $creationType Set to DataObject::CREATE_OBJECT, DataObject::CREATE_HYDRATED,
+     *   DataObject::CREATE_MEMORY_HYDRATED or DataObject::CREATE_SINGLETON. Used by Silverstripe internals as best
```
Comment (madmatt) [inline src/ORM/DataObject.php:349]: The first occurrence of `as` should be `and` here.

File: tests/php/ORM/DataObjectTest.php [hunk 3]
```diff
@@ -1669,6 +1669,31 @@ public function testNewClassInstance()
...
+        $this->assertEquals($changedFields['ClassName']['before'], DataObjectTest\Team::class);
+        $this->assertEquals($changedFields['ClassName']['after'], DataObjectTest\SubTeam::class);
+        $this->assertEquals($changedFields['RecordClassName']['before'], DataObjectTest\Team::class);
+        $this->assertEquals($changedFields['RecordClassName']['after'], DataObjectTest\SubTeam::class);
+
+        $changedDO->write();
+
+        $this->assertInstanceOf(DataObjectTest\SubTeam::class, $changedDO);
+        $this->assertEquals($changedDO->ClassName, DataObjectTest\SubTeam::class);
+        $this->assertNotEmpty($changedDO->ID, );
```
Comment (maxime-rainville) [inline tests/php/ORM/DataObjectTest.php:1694]:
```suggestion
        $this->assertNotEmpty(
            $changedDO->ID,
            'An ID has been generated for the new class instance on save'
        );
```

**Comment by maxime-rainville** [comment]
This didn't make the cut off for the 4.7 release. We're thinking we might be able to call it a bug fix since it's technically not a an API change. This would allow us to ship it in 4.7.1. We'll sort it out, once 4.7.0 stable is out the door.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
There's a lot of duplicated code between the regular hydration and the memory hydration. If we do a bugfix sometime in the future on one of them we might forget to fix the other one. Should create a private method with shared code.

**Comment by sminnee** [comment]
Or use have both options branch to the same case block, and do a further conditional/ternary inside that for the 1 difference.

**Comment by maxime-rainville** [comment]
Addressed @emteknetnz's feedback. I moved the hydration logic into a private `hydrate` method with a flag that account for the difference between CREATE_HYDRATED and CREATE_MEMORY_HYDRATED.

**Comment by KhushbuFuletra** [comment]
Hi, I am waiting for this one. 
when do you think you will be able to tag this PR?

**Comment by KhushbuFuletra** [comment]
> It's going to be in the 4.8.0 release which is still a couple months away.

Thanks, @maxime-rainville  for the heads-up.

**Pull request merged**